File: scripts/export_seurat_wnn_golden.R
#!/usr/bin/env Rscript
# Produces the Seurat side of tests/seurat_wnn_5_5_1_golden.json.
# Driven by scripts/export_seurat_wnn_golden.py, which supplies the shared
# embeddings and neighbour indices and consumes the tables written here.
#
# The matched tables come from Seurat's own routines applied to Scarf's
# candidate pool and bandwidth: L2Norm, PredictAssay, impute_dist and NNdist do
# the arithmetic, and the kernel, score and softmax expressions are copied from
# the body of FindModalityWeights. The default tables come from the public
# FindMultiModalNeighbors with its shipped defaults, which Scarf does not try to
# reproduce.

suppressPackageStartupMessages({
  library(Seurat)
  library(SeuratObject)
})

args <- commandArgs(trailingOnly = TRUE)
if (length(args) != 1L) {
  stop("usage: export_seurat_wnn_golden.R <work-dir>")
}
work <- args[[1]]

read_matrix <- function(name) {
  as.matrix(read.table(file.path(work, name), header = FALSE, colClasses = "numeric"))
}

write_matrix <- function(values, name) {
  values <- as.matrix(values)
  lines <- apply(values, 1, function(row) paste(sprintf("%.17g", row), collapse = "\t"))
  writeLines(lines, file.path(work, name))
}

spec <- read.table(
  file.path(work, "modalities.tsv"),
  header = FALSE,
  sep = "\t",
  col.names = c("name", "stem", "reduction"),
  stringsAsFactors = FALSE
)
if (nrow(spec) < 2L || anyDuplicated(spec$name) || anyDuplicated(spec$reduction)) {
  stop("modality specification must contain at least two unique entries")
}
modality_names <- spec$name
reductions <- spec$reduction
embeddings <- setNames(lapply(seq_len(nrow(spec)), function(i) {
  read_matrix(paste0(spec$stem[[i]], "_embedding.tsv"))
}), reductions)
# Python writes zero-based, self-free neighbour rows.
indices <- setNames(lapply(seq_len(nrow(spec)), function(i) {
  read_matrix(paste0(spec$stem[[i]], "_indices.tsv")) + 1
}), reductions)

n_cells <- nrow(embeddings[[1]])
k <- ncol(indices[[1]])
if (
  any(vapply(embeddings, nrow, integer(1)) != n_cells) ||
  any(vapply(indices, nrow, integer(1)) != n_cells) ||
  any(vapply(indices, ncol, integer(1)) != k)
) {
  stop("all modalities must use the same cell count and neighbour count")
}
cell_names <- paste0("cell", seq_len(n_cells))
for (r in reductions) {
  rownames(embeddings[[r]]) <- cell_names
  colnames(embeddings[[r]]) <- paste0(toupper(r), "_", seq_len(ncol(embeddings[[r]])))
}

counts <- matrix(
  1,
  nrow = 5,
  ncol = n_cells,
  dimnames = list(paste0("gene", 1:5), cell_names)
)
object <- CreateSeuratObject(counts = counts, assay = "RNA")
dims <- setNames(lapply(embeddings, function(values) seq_len(ncol(values))), reductions)
for (r in reductions) {
  object[[r]] <- CreateDimReducObject(
    embeddings = embeddings[[r]],
    key = paste0(toupper(r), "_"),
    assay = "RNA"
  )
}

# Seurat's own row normalization, matching l2_normalize=True on the Scarf side.
normed <- lapply(reductions, function(r) {
  Seurat:::L2Norm(mat = Embeddings(object, r)[, dims[[r]], drop = FALSE])
})
names(normed) <- reductions
for (r in reductions) {
  object[[paste0(r, ".norm")]] <- CreateDimReducObject(
    embeddings = normed[[r]],
    key = paste0("norm", Key(object[[r]])),
    assay = DefaultAssay(object[[r]])
  )
}

# nearest.dist is the closest non-self neighbour and sigma is the k-th non-self
# neighbour minus that, which is the bandwidth Scarf uses. Seurat reaches the
# same two quantities as Distances(nn)[, 2] and Distances(nn)[, sigma.idx] on a
# neighbour object whose first column is the cell itself.
own_distances <- lapply(reductions, function(r) {
  embedding <- normed[[r]]
  t(sapply(seq_len(n_cells), function(i) {
    neighbours <- embedding[indices[[r]][i, ], , drop = FALSE]
    sqrt(rowSums(sweep(neighbours, 2, embedding[i, ], FUN = "-")^2))
  }))
})
names(own_distances) <- reductions
nearest_dist <- lapply(own_distances, function(d) apply(d, 1, min))
sigma <- lapply(reductions, function(r) {
  apply(own_distances[[r]], 1, function(row) sort(row)[k]) - nearest_dist[[r]]
})
names(sigma) <- reductions

within_impute <- lapply(reductions, function(r) {
  PredictAssay(
    object = object,
    nn.idx = indices[[r]],
    assay = DefaultAssay(object[[r]]),
    reduction = paste0(r, ".norm"),
    dims = dims[[r]],
    return.assay = FALSE,
    verbose = FALSE
  )
})
names(within_impute) <- reductions
cross_impute <- lapply(reductions, function(r) {
  sources <- setdiff(reductions, r)
  values <- lapply(sources, function(source) {
    PredictAssay(
      object = object,
      nn.idx = indices[[source]],
      assay = DefaultAssay(object[[r]]),
      reduction = paste0(r, ".norm"),
      dims = dims[[r]],
      return.assay = FALSE,
      verbose = FALSE
    )
  })
  setNames(values, sources)
})
names(cross_impute) <- reductions

within_dist <- lapply(reductions, function(r) {
  Seurat:::impute_dist(
    x = normed[[r]],
    y = t(within_impute[[r]]),
    nearest.dist = nearest_dist[[r]]
  )
})
cross_dist <- lapply(reductions, function(r) {
  lapply(cross_impute[[r]], function(values) {
    Seurat:::impute_dist(
      x = normed[[r]],
      y = t(values),
      nearest.dist = nearest_dist[[r]]
    )
  })
})
names(within_dist) <- names(cross_dist) <- reductions

# The next expressions follow the N-modality loops in FindModalityWeights.
within_kernel <- lapply(reductions, function(r) exp(-1 * (within_dist[[r]] / sigma[[r]])))
cross_kernel <- lapply(reductions, function(r) {
  lapply(cross_dist[[r]], function(values) exp(-1 * (values / sigma[[r]])))
})
names(within_kernel) <- names(cross_kernel) <- reductions
score <- lapply(reductions, function(r) {
  lapply(cross_kernel[[r]], function(values) {
    MinMax(data = within_kernel[[r]] / (values + 1e-04), min = 0, max = 200)
  })
})
names(score) <- reductions
score_columns <- unlist(score, recursive = FALSE, use.names = FALSE)
score_total <- rowSums(exp(do.call(cbind, score_columns)))
weights <- lapply(reductions, function(r) {
  rowSums(exp(do.call(cbind, score[[r]]))) / score_total
})
names(weights) <- reductions

# Blend the modality kernels over the union pool and keep the k strongest, breaking
# ties by ascending cell index the way Scarf's lexsort does.
matched_indices <- matrix(0L, nrow = n_cells, ncol = k)
matched_affinity <- matrix(0, nrow = n_cells, ncol = k)
for (i in seq_len(n_cells)) {
  pool <- sort(unique(unlist(lapply(indices, function(values) values[i, ]))))
  affinity <- rep(0, length(pool))
  for (r in reductions) {
    adjusted <- Seurat:::NNdist(
      nn.idx = matrix(pool, nrow = 1),
      embeddings = normed[[r]],
      query.embeddings = normed[[r]][i, , drop = FALSE],
      nearest.dist = nearest_dist[[r]][i]
    )[[1]]
    affinity <- affinity + weights[[r]][i] * exp(-1 * (adjusted / sigma[[r]][i]))
  }
  chosen <- order(-affinity, pool)[seq_len(k)]
  matched_indices[i, ] <- pool[chosen] - 1L
  matched_affinity[i, ] <- affinity[chosen]
}

write_matrix(matched_indices, "matched_indices.tsv")
write_matrix(matched_affinity, "matched_affinities.tsv")
write_matrix(do.call(cbind, weights), "matched_modality_weights.tsv")

# Public defaults. Scarf deliberately differs here, so this table only supports
# a drift check on how far apart the two stay.
weight_names <- paste0(reductions, ".weight")
default_object <- FindMultiModalNeighbors(
  object = object,
  reduction.list = as.list(reductions),
  dims.list = unname(dims[reductions]),
  k.nn = k,
  knn.range = min(200L, n_cells - 1L),
  modality.weight.name = weight_names,
  verbose = FALSE
)
default_nn <- default_object[["weighted.nn"]]
default_indices <- Indices(default_nn)
write_matrix(default_indices - 1L, "default_indices.tsv")
write_matrix(
  default_object[[]][, weight_names, drop = FALSE],
  "default_modality_weights.tsv"
)

writeLines(
  c(
    paste0("seurat\t", as.character(packageVersion("Seurat"))),
    paste0("seuratObject\t", as.character(packageVersion("SeuratObject"))),
    paste0("r\t", paste(R.version$major, R.version$minor, sep = ".")),
    paste0("kNn\t", k),
    paste0("nCells\t", n_cells)
  ),
  file.path(work, "versions.tsv")
)

File: scripts/export_seurat_wnn_golden.py
#!/usr/bin/env python
"""Regenerate the two-way or synthetic three-way Seurat WNN golden fixture.

The default mode loads explicitly named reduction artifacts from the public
CITE-seq store. Pass one ``--reduction ASSAY=ARTIFACT_ID`` for every assay. The
``--synthetic-three-way`` mode creates deterministic RNA, ATAC, and ADT
embeddings without an external dataset. Both modes compute exact self-free
neighbour rows and hand an arbitrary ordered modality list to the R companion.

Needs an R installation carrying Seurat. Point --rscript at it, for example a
conda environment created with `conda create -p /tmp/scarf-seurat-env
-c conda-forge r-base r-seurat`.
"""

import argparse
import json
import subprocess
import tempfile
from pathlib import Path

import numpy as np

from scarf.storage import ArtifactRef
from scarf.storage.types import as_zarr_array

REPO = Path(__file__).resolve().parent.parent
DEFAULT_STORE = (
    REPO / "docs/source/tutorials/scarf_datasets/tenx_8K_pbmc_citeseq/data.zarr"
)
DEFAULT_FIXTURE = REPO / "tests/seurat_wnn_5_5_1_golden.json"
DEFAULT_THREE_WAY_FIXTURE = REPO / "tests/seurat_wnn_3way_5_5_1_golden.json"
R_SCRIPT = Path(__file__).resolve().parent / "export_seurat_wnn_golden.R"


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--store", type=Path, default=DEFAULT_STORE)
    parser.add_argument("--rscript", type=Path, default=Path("Rscript"))
    parser.add_argument("--output", type=Path)
    parser.add_argument("--assays", nargs="+", default=["RNA", "ADT"])
    parser.add_argument(
        "--reduction",
        action="append",
        default=[],
        metavar="ASSAY=ARTIFACT_ID",
        help="Explicit assay-scoped reduction artifact (repeat per assay)",
    )
    parser.add_argument("--synthetic-three-way", action="store_true")
    parser.add_argument("--cells", type=int)
    parser.add_argument("--k", type=int, default=20)
    parser.add_argument("--seed", type=int, default=20210601)
    return parser.parse_args()


def parse_reduction_refs(
    values: list[str],
    assays: tuple[str, ...],
) -> dict[str, ArtifactRef]:
    refs: dict[str, ArtifactRef] = {}
    for value in values:
        assay, separator, artifact_id = value.partition("=")
        if not separator or not assay or not artifact_id:
            raise ValueError("--reduction values must use ASSAY=ARTIFACT_ID")
        if assay in refs:
            raise ValueError(f"Duplicate reduction for assay {assay!r}")
        refs[assay] = ArtifactRef(
            scope="assay",
            assay=assay,
            kind="reduction",
            artifact_id=artifact_id,
        )
    expected = set(assays)
    actual = set(refs)
    if actual != expected:
        missing = sorted(expected - actual)
        unexpected = sorted(actual - expected)
        raise ValueError(
            "Explicit reductions must match --assays exactly; "
            f"missing={missing!r}, unexpected={unexpected!r}"
        )
    return refs


def load_embeddings(
    store: Path,
    reductions: dict[str, ArtifactRef],
) -> dict[str, np.ndarray]:
    import scarf

    scarf.configure_output(level="ERROR", progress=False)
    assays = tuple(reductions)
    datastore = scarf.DataStore(str(store), default_assay=assays[0], nthreads=4)
    embeddings = {}
    for assay, reference in reductions.items():
        group = datastore.load_artifact(reference)
        embeddings[assay] = np.asarray(as_zarr_array(group["data"])[:])
    return embeddings


def synthetic_three_way_embeddings(
    n_cells: int,
    seed: int,
) -> dict[str, np.ndarray]:
    generator = np.random.default_rng(seed)
    latent = generator.normal(size=(n_cells, 4))
    return {
        "RNA": latent @ generator.normal(size=(4, 6))
        + generator.normal(scale=0.08, size=(n_cells, 6)),
        "ATAC": latent @ generator.normal(size=(4, 5))
        + generator.normal(scale=0.12, size=(n_cells, 5)),
        "ADT": latent @ generator.normal(size=(4, 3))
        + generator.normal(scale=0.05, size=(n_cells, 3)),
    }


def exact_neighbors(embedding: np.ndarray, k: int) -> np.ndarray:
    """Self-free neighbour rows ordered by ascending float64 distance."""
    gram = embedding @ embedding.T
    square = np.diag(gram)
    distances = np.sqrt(np.maximum(square[:, None] - 2.0 * gram + square[None, :], 0.0))
    np.fill_diagonal(distances, np.inf)
    return np.argsort(distances, axis=1, kind="stable")[:, :k].astype(np.uint32)


def write_table(path: Path, values: np.ndarray) -> None:
    fmt = "%d" if np.issubdtype(values.dtype, np.integer) else "%.17g"
    np.savetxt(path, np.atleast_2d(values), fmt=fmt, delimiter="\t")


def read_table(path: Path) -> np.ndarray:
    return np.atleast_2d(np.loadtxt(path, dtype=np.float64, delimiter="\t"))


def main() -> None:
    args = parse_args()
    reduction_sources: dict[str, dict[str, str | None]] | None = None
    if args.synthetic_three_way:
        if args.reduction:
            raise ValueError("--reduction is not used with --synthetic-three-way")
        modality_names = ("RNA", "ATAC", "ADT")
        n_cells = args.cells or 48
        selected_embeddings = synthetic_three_way_embeddings(n_cells, args.seed)
        dataset = "synthetic_three_modality"
        dataset_source = "scripts/export_seurat_wnn_golden.py"
        output = args.output or DEFAULT_THREE_WAY_FIXTURE
    else:
        modality_names = tuple(args.assays)
        if len(modality_names) < 2:
            raise ValueError("Golden generation requires at least two assays")
        if len(set(modality_names)) != len(modality_names):
            raise ValueError("Golden generation requires unique assay names")
        if modality_names != ("RNA", "ADT") and args.output is None:
            raise ValueError("--output is required for a non-default assay list")
        reductions = parse_reduction_refs(args.reduction, modality_names)
        embeddings = load_embeddings(args.store, reductions)
        generator = np.random.default_rng(args.seed)
        total = embeddings[modality_names[0]].shape[0]
        n_cells = args.cells or 300
        if n_cells > total:
            raise ValueError(f"Store holds {total} cells, cannot sample {n_cells}")
        selected = np.sort(generator.choice(total, size=n_cells, replace=False))
        selected_embeddings = {
            name: np.asarray(embeddings[name][selected], dtype=np.float64)
            for name in modality_names
        }
        dataset = args.store.parent.name
        dataset_source = (
            "scarf.cytebase.connect('scarf_docs')"
            if args.store == DEFAULT_STORE
            else str(args.store)
        )
        reduction_sources = {
            name: reductions[name].to_dict() for name in modality_names
        }
        output = args.output or DEFAULT_FIXTURE
    if args.k >= n_cells:
        raise ValueError("k must be smaller than the number of fixture cells")

    # Storage values are widened to float64 so R and Python start from
    # identical numbers and recompute identical distances.
    selected_embeddings = {
        name: np.asarray(selected_embeddings[name], dtype=np.float64)
        for name in modality_names
    }
    neighbor_indices = {
        name: exact_neighbors(selected_embeddings[name], args.k)
        for name in modality_names
    }

    with tempfile.TemporaryDirectory(prefix="scarf-wnn-golden-") as raw_work:
        work = Path(raw_work)
        modality_specs = []
        for index, name in enumerate(modality_names):
            stem = f"modality_{index + 1}"
            reduction = f"mod{index + 1}"
            write_table(work / f"{stem}_embedding.tsv", selected_embeddings[name])
            write_table(work / f"{stem}_indices.tsv", neighbor_indices[name])
            modality_specs.append(f"{name}\t{stem}\t{reduction}")
        (work / "modalities.tsv").write_text("\n".join(modality_specs) + "\n")
        subprocess.run(
            [str(args.rscript), str(R_SCRIPT), str(work)],
            check=True,
        )
        versions = dict(
            line.split("\t")
            for line in (work / "versions.tsv").read_text().splitlines()
        )
        matched_indices = read_table(work / "matched_indices.tsv")
        matched_affinities = read_table(work / "matched_affinities.tsv")
        matched_weights = read_table(work / "matched_modality_weights.tsv")
        default_indices = read_table(work / "default_indices.tsv")
        default_weights = read_table(work / "default_modality_weights.tsv")

    provenance = {
        "package": "Seurat",
        "packageVersion": versions["seurat"],
        "seuratObjectVersion": versions["seuratObject"],
        "rVersion": versions["r"],
        "sourceRepository": "https://github.com/satijalab/seurat",
        "dataset": dataset,
        "datasetSource": dataset_source,
        "generator": "scripts/export_seurat_wnn_golden.py",
        "nCells": n_cells,
        "kNn": int(args.k),
        "cellSampleSeed": int(args.seed),
        "modalityNames": list(modality_names),
        "l2Normalize": True,
        "matchedFunctions": [
            "Seurat:::L2Norm",
            "Seurat:::PredictAssay",
            "Seurat:::impute_dist",
            "Seurat:::NNdist",
            "Seurat::MinMax",
        ],
        "matchedNote": (
            "Seurat routines evaluated on Scarf's union candidate pool and "
            "nearest-to-k-th non-self distance-span bandwidth. The kernel, "
            "score and softmax expressions are copied from the body of "
            "FindModalityWeights."
        ),
        "defaultFunction": "Seurat::FindMultiModalNeighbors",
        "defaultNote": (
            "Shipped defaults, including the wider knn.range search and the "
            "SNN-far bandwidth. Scarf does not reproduce these, so the "
            "comparison is statistical."
        ),
    }
    if reduction_sources is not None:
        provenance["reductions"] = reduction_sources

    fixture = {
        "provenance": provenance,
        "matched": {
            "modalityWeights": matched_weights.tolist(),
            "neighborIndices": matched_indices.astype(int).tolist(),
            "neighborAffinities": matched_affinities.tolist(),
        },
        "seuratDefault": {
            "modalityWeights": default_weights.tolist(),
            "neighborIndices": default_indices.astype(int).tolist(),
        },
    }
    if modality_names == ("RNA", "ADT") and not args.synthetic_three_way:
        fixture["inputs"] = {
            "rnaEmbedding": selected_embeddings["RNA"].tolist(),
            "adtEmbedding": selected_embeddings["ADT"].tolist(),
            "rnaIndices": neighbor_indices["RNA"].astype(int).tolist(),
            "adtIndices": neighbor_indices["ADT"].astype(int).tolist(),
        }
    else:
        fixture["inputs"] = {
            "modalityNames": list(modality_names),
            "embeddings": {
                name: selected_embeddings[name].tolist() for name in modality_names
            },
            "neighborIndices": {
                name: neighbor_indices[name].astype(int).tolist()
                for name in modality_names
            },
        }
    output.write_text(json.dumps(fixture, indent=1) + "\n")
    print(f"wrote {output} ({output.stat().st_size / 1024:.0f} KB)")


if __name__ == "__main__":
    main()
